### PR TITLE
Default auto_axes not over manual axes

### PR DESCRIPTION
**Pull request by oliverdutton-iso**
Alter auto_axes so that when axes=None in the function call it only acts over the explicit axes. Allowing examples like below, where high level shard_maps are used.

```python
# !pip install -U --pre jax jaxlib -i https://us-python.pkg.dev/ml-oss-artifacts-published/jax/simple/

import functools

import jax
import jax.numpy as jnp

from jax.experimental import mesh_utils
from jax.sharding import Mesh, PartitionSpec as P

@functools.partial(jax.sharding.auto_axes, out_sharding=P()) # as expected, axes=('model',) does fix it
def simple_func(inputs):
    # simple gather
    return inputs[(0,0), (1,1)]

device_mesh = mesh_utils.create_device_mesh(mesh_shape=(1,1))
mesh = Mesh(
    device_mesh,
    axis_names=("data", "model"),
    axis_types=(jax.sharding.AxisType.Explicit, jax.sharding.AxisType.Explicit)
)
func = jax.shard_map(
    simple_func,
    out_specs=P("data"),
    axis_names={'data'},
    check_vma=False,
)
func = jax.jit(func)

with jax.sharding.use_mesh(mesh):
  inputs = jnp.zeros((16, 16, 16), out_sharding=P('data'))
  out = func(inputs).block_until_ready()
```

### CODE REVIEW

**Comment by yashk2810** [comment]
Seems reasonable but one thing I worry about is this might cause future problems where we might want to allow `auto_axes` to indeed convert Manual to Auto.

But if `axes=None` means only explicit axes, then doing the above change would be difficult in the future. 

Why not just mention the axis you want instead as you showed?

**Comment by oliverdutton-iso** [comment]
The auto_axes usages can happen in the JAX codebase internals
or possibly even in a transpose of function which can be awkward

e.g. https://github.com/jax-ml/jax/blob/0f1d274e0c439a888277b0390964a64f54512af2/jax/_src/numpy/lax_numpy.py#L6371-L6379
and  https://github.com/jax-ml/jax/blob/0f1d274e0c439a888277b0390964a64f54512af2/jax/_src/random.py#L389-L390 

`auto_axes` can still be forcibly requested to convert manual to auto, just I think that behavior should be opt-in rather than by default

**Comment by oliverdutton-iso** [comment]
see #30003
